cmake/NeuronTestHelper.cmake

```diff
@@ -0,0 +1,361 @@
+    add_custom_command(
+      OUTPUT ${modfile_build_paths}
+      DEPENDS ${modfiles}
+      COMMAND ${CMAKE_COMMAND} -E copy_if_different ${modfiles} "${nrnivmodl_working_directory}"
+      COMMENT "Copying modfiles needed for test ${NRN_ADD_TEST_GROUP}::${NRN_ADD_TEST_NAME}")
+    # Construct the names of the important output files
+    set(special "${nrnivmodl_working_directory}/${CMAKE_HOST_SYSTEM_PROCESSOR}/special")
+    # Add the custom command to generate the binaries. Get nrnivmodl from the build directory. At
+    # the moment it seems that `nrnivmodl` is generated at configure time, so there is no target to
+    # depend on and it should always be available, but it will try and link against libnrniv.so and
+    # libcoreneuron.so so we must depend on those. TODO: could the logic of `nrnivmodl` be
+    # translated to CMake, so it can be called natively here and the `nrnivmodl` executable would be
+    # a wrapper that invokes CMake?
+    set(output_binaries "${special}")
+    set(nrnivmodl_dependencies nrniv_lib)
+    if(requires_coreneuron)
+      list(APPEND output_binaries "${special}-core")
+      list(APPEND nrnivmodl_dependencies coreneuron)
+    endif()
+    add_custom_command(
+      OUTPUT ${output_binaries}
+      DEPENDS ${nrnivmodl_dependencies} ${modfile_build_paths}
+      COMMAND ${CMAKE_COMMAND} -E env ${NRN_TEST_ENV} PATH=${CMAKE_BINARY_DIR}/bin:$ENV{PATH}
+              ${nrnivmodl_command}
+      COMMENT "Building special[-core] for test ${NRN_ADD_TEST_GROUP}::${NRN_ADD_TEST_NAME}"
+      WORKING_DIRECTORY ${nrnivmodl_working_directory})
+    # Add a target that depends on the binaries
+    add_custom_target(${binary_target_name} DEPENDS ${output_binaries})
+    # Make the test-group-level target depend on this new target.
+    add_dependencies(${prefix} ${binary_target_name})
+  endif()
+  # Set up the actual test. First, collect the script files that need to be copied into the test-
+  # specific working directory and copy them there.
+  file(MAKE_DIRECTORY "${working_directory}")
+  foreach(script_pattern ${script_patterns})
+    file(GLOB script_files "${test_source_directory}/${script_pattern}")
+    add_custom_command(
+      TARGET ${prefix} POST_BUILD
+      COMMAND ${CMAKE_COMMAND} -E copy_if_different ${script_files} "${working_directory}"
+      COMMENT "Copying scripts needed to run test ${NRN_ADD_TEST_GROUP}::${NRN_ADD_TEST_NAME}")
+  endforeach()
+  # Construct the name of the test and store it in a parent-scope list to be used when setting up
+  # the comparison job
+  set(test_name "${NRN_ADD_TEST_GROUP}::${NRN_ADD_TEST_NAME}")
+  set(group_members "${${prefix}_TESTS}")
+  list(APPEND group_members "${test_name}")
+  set(${prefix}_TESTS
+      "${group_members}"
+      PARENT_SCOPE)
+  # Add the actual test job, including the `special` and `special-core` binaries in the path. TODO:
+  # do we need to manipulate PYTHONPATH too to make `python options.py` invocations work?
```

Comment:
PYTHONPATH is required for running tests from build directory  but I didn't understand `options.py` part here.

Reply:
I think the point here was that for `special` we manipulate `$PATH` so that at the call site of `nrn_add_test` you can just write:
```cmake
  nrn_add_test(NAME foo COMMAND special ...)
```
instead of
```cmake
  nrn_add_test(NAME foo COMMAND ${some_path_to_special} ...)
```
and the logic around what directories different commands (`nrnivmodl`, the test commands, ...) are run in is confined to `NeuronTestHelper.cmake`. In fact the `$PYTHONPATH` [is now set](https://github.com/neuronsimulator/nrn/blob/79c9965e9750b426702b9bfdc08c67f8b55940a0/test/CMakeLists.txt#L32), but I wasn't 100% sure if there will be problems if a test script does something more complicated (e.g. shipping multiple python files, importing one from the other etc.)

Reply:
for PYTHONPATH:
* `PYTHONPATH=${PROJECT_BINARY_DIR}/lib/python:${PROJECT_SOURCE_DIR}/test/rxd:$ENV{PYTHONPATH})` could be considered as default, project level path (`test/rxd` could be removed but we can assume it's given)
* after this, each test can have it's own **additional** `PYTHON_LIBRARY_PATH` and `HOC_LIBRARY_PATH`.

Reply:
Could you elaborate on what you mean with `PYTHON_LIBRARY_PATH` here please?

Reply:
Sorry, Just typo! I meant there could be a directory where we have to set `PYTHONPATH`.

Reply:
OK. I suggest that we just leave it as is for now. If we need a special way of adding extra python paths then it can be added later, maybe it will never turn out to be necessary.